clients/venice-push-job/src/main/java/com/linkedin/venice/hadoop/input/kafka/chunk/ChunkAssembler.java

```diff
@@ -24,129 +20,156 @@
-    for (byte[] chunkBytes: allChunkBytes) {
-      if (chunksFound == valueChunks.length) {
-        break;
+      if (reusedMapperValue.schemaId > 0) {
+        if (latestChunkedValueManifest != null) {
+          // Ignore older entries since a more recent manifest is discovered.
+          continue;
+        }
+        /**
+         * The latest event is a put event.
+         */
+        return new ValueBytesAndSchemaId(
+            ByteUtils.extractByteArray(reusedMapperValue.value),
+            reusedMapperValue.schemaId,
+            reusedMapperValue.replicationMetadataVersionId,
+            reusedMapperValue.replicationMetadataPayload);
       }
-      chunk = KAFKA_INPUT_MAPPER_VALUE_AVRO_SPECIFIC_DESERIALIZER.deserialize(chunk, chunkBytes);
-      for (int i = 0; i < chunkKeySuffixes.length; i++) {
-        byteBuffer = chunkKeySuffixes[i];
-        if (byteBuffer.equals(chunk.chunkedKeySuffix)) {
-          byte[] valueChunk = new byte[chunk.value.remaining()];
-          totalByteCount += valueChunk.length;
-          chunk.value.get(valueChunk);
-          valueChunks[i] = valueChunk;
-          chunksFound++;
+
+      if (reusedMapperValue.schemaId == AvroProtocolDefinition.CHUNKED_VALUE_MANIFEST.getCurrentProtocolVersion()) {
+        // Only capture the latest manifest
+        if (latestChunkedValueManifest == null) {
+          latestChunkedValueManifestRMDVersionId = reusedMapperValue.replicationMetadataVersionId;
+          latestChunkedValueManifestRMDPayload =
+              ByteBuffer.wrap(ByteUtils.copyByteArray(reusedMapperValue.replicationMetadataPayload));
+          latestChunkedValueManifest = manifestSerializer.deserialize(
+              ByteUtils.extractByteArray(reusedMapperValue.value),
+              AvroProtocolDefinition.CHUNKED_VALUE_MANIFEST.getCurrentProtocolVersion());
+          int chunkCount = latestChunkedValueManifest.keysWithChunkIdSuffix.size();
+          valueChunks = new byte[chunkCount][];
+          chunkKeySuffixes = new ByteBuffer[chunkCount];
+
+          for (int i = 0; i < latestChunkedValueManifest.keysWithChunkIdSuffix.size(); i++) {
+            ByteBuffer byteBuffer = latestChunkedValueManifest.keysWithChunkIdSuffix.get(i);
+            int startPosition = byteBuffer.position() + keyBytes.length;
+            int suffixLength = byteBuffer.remaining() - keyBytes.length;
+            chunkKeySuffixes[i] = ByteBuffer.wrap(byteBuffer.array(), startPosition, suffixLength);
+          }
+        }
+      } else if (reusedMapperValue.schemaId == AvroProtocolDefinition.CHUNK.getCurrentProtocolVersion()) {
+        if (latestChunkedValueManifest == null) {
+          /**
+           * It is possible that KafkaInputFormat could only read the partial chunks for a large message, which means
+           * there are chunks without the corresponding manifest.
+           */
+          continue;
+        }
+        // Collecting chunks
+        for (int i = 0; i < chunkKeySuffixes.length; i++) {
```

Comment:
Just to confirm this is for safe guard or it is possible that chunk can happen out of order?
Cuz I am thinking ideally it should be manifest, chunkN, chunkN-1,...Chunk1 and we don't need to scan this every time for every chunk?

Reply:
yes, this one is slightly inefficient, and it is mostly for the developer to easily understand the code since it is simpler.
Also the inefficiency is kind of minimal since once all the chunks are found, we will break the loop, which means for most of the chunked messages, the run time will be much shorter than before.
It is possible to make it more efficient such as adjusting the `chunKeySuffixes` on the fly to ignore the found chunks, but I feel the complexity is not quite necessary considering the code maintainability.

Reply:
Sure, I believe we can revisit this part later upon future optimization